File: Level 3/doomsday-fuel/solution.py
# This solution hasn't been battle-tested. 
# I simply re-wrote my Java code in Python for fun.

def solution(m):
  terminal_states_list = []
  for a in range(0, len(m)):
    rowASum = sum(m[a])
    if rowASum == 0:
      terminal_states_list.append(a)
      continue
    rowASum -= m[a][a]
    m[a][a] = 0
    for b in range(0, len(m)):
      if m[b][a] == 0: continue
      for c in range(0, len(m)):
        if c == a: continue
        m[b][c] *= rowASum
        m[b][c] += m[a][c] * m[b][a]
      m[b][a] = 0
      simplify(m[b])
  result = []
  for item in terminal_states_list:
    result.append(m[0][item])
  result.append(sum(m[0]))
  if result[-1] == 0:
    result[0] = 1
    result[-1] = 1
  return result

def simplify(m):
  gcd_curr = m[0]
  for i in range(1, len(m)):
    gcd_curr = gcd(gcd_curr, m[i])
  for i in range(0, len(m)):
    m[i] /= gcd_curr

def gcd(a, b): return a if b == 0 else gcd(b, a % b)

File: Level 3/doomsday-fuel/Solution.java
import java.util.ArrayList;

public class Solution {
    public static int[] solution(int[][] m) {
        ArrayList<Integer> terminalStatesList = new ArrayList<>();
        for (int a = 0; a < m.length; a++) {
            int rowASum = rowSum(m[a]);
            if (rowASum == 0) {
                terminalStatesList.add(a);
                continue;
            }
            rowASum -= m[a][a];
            m[a][a] = 0;
            for (int b = 0; b < m.length; b++) {
                if (m[b][a] == 0) continue;
                for (int c = 0; c < m.length; c++) {
                    if (c == a) continue;
                    m[b][c] *= rowASum;
                    m[b][c] += m[a][c] * m[b][a];
                }
                m[b][a] = 0;
                simplify(m[b]);
            }
        }
        int[] result = new int[terminalStatesList.size() + 1];
        for (int i = 0; i < terminalStatesList.size(); i++) {
            result[i] = m[0][terminalStatesList.get(i)];
        }
        result[result.length - 1] = rowSum(m[0]);
        if (result[result.length - 1] == 0) {
            result[0] = 1;
            result[result.length - 1] = 1;
        }
        return result;
    }
    
    private static int rowSum(int[] m) {
        int total = 0;
        for (int item : m) {
            total += item;
        }
        return total;
    }

    private static void simplify(int[] m) {
        int result = m[0];
        for (int i = 1; i < m.length; i++) {
            result = gcd(result, m[i]);
        }
        for (int i = 0; i < m.length; i++) {
            m[i] /= result;
        }
    }

    private static int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}